### ISSUE TITLE
Exception "Headers are read-only, response has already started" when refreshing access token

### ISSUE THREAD

**Issue by jerriep** [IdentityModel/IdentityModel.AspNetCore#108]
Whenever the access token is refreshed, I get an exception "Headers are read-only, response has already started"

There is nothing special about my configuration:

```
services.AddAuthentication(options =>
{
    options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultChallengeScheme = "Auth0";
})
.AddCookie(options =>
{
    // ...
})
.AddOpenIdConnect("Auth0", options =>
{
    // ...
});
services.AddAccessTokenManagement();
services.AddUserAccessTokenClient("MyApiClient", httpClient =>
{
    var uri = Configuration.GetValue<string>("MyApi:Url");
    httpClient.BaseAddress = new Uri(uri);
});

```

Here is there relevant parts of the stacktrace

```
 Error: System.InvalidOperationException: Headers are read-only, response has already started.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.ThrowHeadersReadOnlyException()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpHeaders.Microsoft.AspNetCore.Http.IHeaderDictionary.set_Item(String key, StringValues value)
   at Microsoft.AspNetCore.Http.ResponseCookies.Append(String key, String value, CookieOptions options)
   at Microsoft.AspNetCore.Authentication.Cookies.ChunkingCookieManager.AppendResponseCookie(HttpContext context, String key, String value, CookieOptions options)
   at Microsoft.AspNetCore.Authentication.Cookies.CookieAuthenticationHandler.HandleSignInAsync(ClaimsPrincipal user, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationService.SignInAsync(HttpContext context, String scheme, ClaimsPrincipal principal, AuthenticationProperties properties)
   at IdentityModel.AspNetCore.AccessTokenManagement.AuthenticationSessionUserTokenStore.StoreTokenAsync(ClaimsPrincipal user, String accessToken, Int32 expiresIn, String refreshToken)
   at IdentityModel.AspNetCore.AccessTokenManagement.AccessTokenManagementService.RefreshUserAccessTokenAsync()
   at IdentityModel.AspNetCore.AccessTokenManagement.AccessTokenManagementService.<GetUserAccessTokenAsync>b__12_1()
   at IdentityModel.AspNetCore.AccessTokenManagement.AccessTokenManagementService.GetUserAccessTokenAsync(Boolean forceRenewal)
   at Microsoft.AspNetCore.Authentication.TokenManagementHttpContextExtensions.GetUserAccessTokenAsync(HttpContext context, Boolean forceRenewal)
   at IdentityModel.AspNetCore.AccessTokenManagement.UserAccessTokenHandler.SetTokenAsync(HttpRequestMessage request, Boolean forceRenewal)
   at IdentityModel.AspNetCore.AccessTokenManagement.UserAccessTokenHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
```

**Comment by leastprivilege** [comment]
Can you re-produce that with the MVC sample in the repo?

**Comment by jerriep** [comment]
I'll try and create a reproducible sample for you

**Comment by leastprivilege** [comment]
That wasn't my question. Have you tried to run the sample in the repo?

**Comment by jerriep** [comment]
The `TokenManagement3` sample works fine for me. I can log in and call the API endpoint as a user. If I wait a couple of minutes and try again, it seems to refresh the access token correctly.

**Comment by leastprivilege** [comment]
OK - so the question is then what you do different...

**Comment by jerriep** [comment]
I'll try and narrow it down to a point where I can either identify the difference and fix it on my side or at least give you a small reproducible sample.

**Comment by jerriep** [comment]
Ok, so one thing that is vastly different is that the application in question is a Blazor Server Side application. One problem with this scenario is that [HttpContext can be null](https://github.com/dotnet/aspnetcore/issues/12601). 

This causes problems because the `UserAccessTokenHandler` depends on the HttpContext:
https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/cdf9fa5e2f878b7d2c6094a960498eb7d854e688/src/AccessTokenManagement/UserAccessTokenHandler.cs#L55

Everything works fine locally, but when deployed to Azure, that line throws an exception because `_httpContextAccessor.HttpContext` returns null.

Bottom line is we'll have to take a different approach with this, so I'm not even going to waste your or my time further by trying to figure this out.

Apologies for messing you around.

**Comment by pigwing** [comment]
@jerriep Hi,I have encountered the same problem. Can you find a solution

**Comment by leastprivilege** [comment]
@jerriep Did you ever figure out what the problem was?

**Comment by jerriep** [comment]
@leastprivilege No, nothing beyond what I wrote above that it was somehow related to Blazor Server. We decided to manually handle the storing and refreshing of tokens and I soon after moved off that project.

**Comment by brockallen** [comment]
@jerriep If you did the token exchange manually, where did you store the new access token?

**Comment by jerriep** [comment]
I do not recall the details anymore. As a matter of fact, I may even be wrong about how we resolved it. This was more than a year ago and as stated before, I moved on shortly afterwards.
